src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListenerProvider+NullOperationListener.cs

```diff
@@ -19,10 +20,42 @@ private sealed class NullOperationListener : IAsynchronousOperationListener
                 [CallerFilePath] string filePath = "",
                 [CallerLineNumber] int lineNumber = 0) => EmptyAsyncToken.Instance;
 
-            public async Task<bool> Delay(TimeSpan delay, CancellationToken cancellationToken)
+            public Task<bool> Delay(TimeSpan delay, CancellationToken cancellationToken)
             {
-                await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
-                return true;
+                // This could be as simple as:
+                //     await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+                //     return true;
+                // However, whereas in general cancellation is expected to be rare and thus throwing
+                // an exception in response isn't very impactful, here it's expected to be the case
+                // more often than not as the operation is being used to delay an operation because
+                // it's expected something else is going to happen to obviate the need for that
+                // operation.  Thus, we can spend a little more code avoiding the additional throw
+                // for the common case of an exception occurring.
+
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return Task.FromCanceled<bool>(cancellationToken);
+                }
+
+                var t = Task.Delay(delay, cancellationToken);
+                if (t.IsCompleted)
```

Comment:
can we comment when this is expect4ed to happen?  is this effectively just saying that the delay requested is so short as to be effectively no delay at all?

Reply:
if this is for no-delay.  can we just check that itself (i.e. `delay <= 0`) and return in that case?

Reply:
I prefer the current form, since it applies the best known information we have to the same variable `t` which is used in the `ContinueWith` below.

Reply:
It could happen for a 0 delay, or if there's a non-zero delay but things got scheduled in such a way that by the time we check IsCompleted the timer's already fired, or if the cancellationToken has had cancellation requested by now even if it didn't in the fast-path check earlier.  It's hopefully not common.